refactor(datasource): migrate to class based datasource

JamieMagee opened the pull request:
A small experiment into what OOP/class based datasources might look like. Picked Cdnjs as it's the smallest & simplest.

With this approach we can share common logic, like error handling, rate limiting, etc. between different datasources, instead of having to reimplement it each time we write a new datasource. Currently there's nothing shared, as it's only 1 datasource, but the interesting stuff will come with the 2nd datasource!

viceice commented on lib/datasource/index.spec.ts:
```diff
@@ -30,7 +30,7 @@ describe('datasource/index', () => {
     expect(datasource.getDatasourceList()).toBeDefined();
   });
-  it('validates dataource', async () => {
+  it('validates datsource', async () => {
```
```suggestion
  it('validates datasource', async () => {
```
🙃

viceice commented on lib/datasource/cdnjs/index.ts:
```diff
@@ -27,38 +28,42 @@ async function downloadLibrary(library: string): CachePromise<CdnjsResponse> {
-    const releases = assets
-      .filter(({ files }) => files.includes(assetName))
-      .map(({ version, sri }) => ({ version, newDigest: sri[assetName] }));
+export class CdnJs extends Datasource {
```
I like the class based idea, we should export the class as `default` or with a fixed alias, so we can generate the imports as we now do.

See `versioning` for sample. I think we need to undo the dynamic datasource loading to be like the other modules.

JamieMagee replied:
I can export as default no problem.

What was the problem the dynamic generation/loading was trying to solve? I understand the alternative is to maintain the list in `lib/datasource/index.ts` manually, but this list doesn't really change that often.

viceice replied:
it was to do not have conflicts when users adding new versionigs / datasources / managers ...

JamieMagee replied:
Oops. I didn't get a notification for your comment.

Is that such a common issue? I feel like it's a rare problem?

viceice replied:
No, shouldn't be a big problem.

@rarkins can / should we go back to static imports and remove the code generation?

rarkins replied:
One remaining problem with maintaining a list was that it wouldn't support dynamic addition of datasources in future.

viceice replied:
As long as we have the function for returning the full datasource map, we can simply add new datasources.

You can see it in action [here](https://github.com/renovatebot/internal-tools/blob/15413767c5beffaebfe172fe72d1c2063b8ec157/src/utils/datasource/index.ts#L1-L9).
It's used by our internal tools. 🙃

viceice replied:
If we maintain a static list, I prefer to move it to a separate file, as it will be very long.

rarkins replied:
I meant for end users to conveniently add custom datasources. If such custom ones need to be pre-configured in an index file, it makes it harder.

viceice replied:
If a user needs to add custom datasources i would suggest him to use the approach used by our internal-tools.

rarkins commented:
Related:
- #5557 - De-duplicate datasource error handling
- #6757 - Add private/auth awareness to datasource index

Also, I'd been thinking the potentially the datasource index could instantiate a `http` instance for each datasource at startup (being sure to include the `id`) and then pass it as part of parameters to `getReleases` or `getDigest`.

viceice commented:
? review or not review? 🤔

rarkins commented:
> ? review or not review? 🤔

Sorry, requested review of wrong branch

JamieMagee commented:
Rebased on top of current `HEAD`. Reverted `crate` changes as this branch has strayed too far from the existing implementation, and rewriting would be simpler than rebasing.

viceice commented on lib/datasource/clojure/index.ts:
```diff
@@ -1,8 +1,26 @@
+import { Datasource } from '../datasource';
+import { getReleases } from '../maven';
 import { MAVEN_REPO } from '../maven/common';
+import { GetReleasesConfig, ReleaseResult } from '../types';
```
```suggestion
import type { GetReleasesConfig, ReleaseResult } from '../types';
```

viceice commented on lib/datasource/cdnjs/index.ts:
```diff
@@ -24,38 +22,43 @@ interface CdnjsResponse {
-      .filter(({ files }) => files.includes(assetName))
-      .map(({ version, sri }) => ({ version, newDigest: sri[assetName] }));
+export class CdnJs extends Datasource {
+  readonly id = 'cdnjs';
```
can we share this duplicated id somehow?

viceice commented on lib/datasource/clojure/index.ts:
```diff
@@ -1,8 +1,26 @@
 
-export { getReleases } from '../maven';
+export class Clojure extends Datasource {
+  readonly id = 'clojure';
+
+  readonly registryStrategy = 'merge';
+
+  readonly customRegistrySupport = true;
+
+  readonly defaultRegistryUrls = ['https://clojars.org/repo', MAVEN_REPO];
```
do we need those duplicates?

JamieMagee replied:
This is an issue I don't yet have a great solution for. If I look at `id`, it's defined in the `DatasourceApi` interface, and I defined it as `abstract id` in the abstract class `Datasource` so that subclasses are required to implement it. However that means that you need to instantiate `ClojureDatasource` in order to read its id.

Maybe dropping the `abstract` modifier, and just making it `static`? [Here's a TS playground link to highlight the issue](https://www.typescriptlang.org/play?#code/IYIwzgLgTsDGEAJYBthjAgIsCaD2ArlLAKYIDeAUAjQpDgJawJQnAAmeAdsgJ4IN2ALjrQGXAOaUAvpUoo0GAMLI8AKyIlsuMIWJkSADwgku7DNvxFSFarXoQmA9ggC8CAOQp1mjwG4ZOQV0BAAxPCgJPAhLXWsDY1NzLBwrfVtA+W5dZBIAOlUJAAoVH1ZYvVI8wQBKLK4c-MKi8Mjoivjq9hqgA)

JamieMagee replied:
Or just instantiating the class?

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,55 @@
+import { ExternalHostError } from '../types/errors/external-host-error';
+import { Http, HttpError } from '../util/http';
+import {
```
```suggestion
import type {
```

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,55 @@
+
+  registryStrategy: 'first' | 'hunt' | 'merge';
+
+  getId(): string {
+    return this.id;
+  }
```
Do we need this wrapper function? i think we can directly use `this.id`

JamieMagee replied:
It's not possible to use properties directly in the constructor

viceice replied:
Then i think it's better to pass id via constructor and making it readonly in datasource base class

viceice commented on lib/datasource/index.spec.ts:
```diff
@@ -32,9 +32,8 @@ describe(getName(__filename), () => {
-    expect(datasource.getDatasourceList()).toBeDefined();
   });
-  it('validates dataource', () => {
+  it('validates datsource', () => {
```
```suggestion
  it('validates datasource', () => {
```
🤭

JamieMagee replied:
Trading 1 spelling mistake for another :smile:

viceice commented on lib/datasource/index.spec.ts:
```diff
@@ -32,9 +32,8 @@ describe(getName(__filename), () => {
   });
   it('returns datasources', () => {
     expect(datasource.getDatasources()).toBeDefined();
-    expect(datasource.getDatasourceList()).toBeDefined();
```
Why this change?

JamieMagee replied:
I think it was a workaround when I had directly imported the datasources in `index.ts`. Reverted.

viceice commented on lib/datasource/index.ts:
```diff
@@ -27,7 +29,10 @@ export const getDatasourceList = (): string[] => Array.from(datasources.keys());
 const cacheNamespace = 'datasource-releases';
 
-function load(datasource: string): DatasourceApi {
+datasources.set('cdnjs', new CdnJs());
+datasources.set('clojure', new Clojure());
```
```suggestion
```

already done in `./api.ts`. wrong merge?

JamieMagee replied:
Yes, I think this was done before before the dynamic to static migration.

JamieMagee replied:
Yes, I think this was done before the dynamic to static migration

viceice commented on lib/datasource/pod/index.ts:
```diff
@@ -42,7 +42,7 @@ function releasesGithubUrl(
   const errorData = { lookupName, err };
 
-  const statusCode = err.response?.statusCode;
+  const statusCode = err.statusCode;
```
revert `err.statusCode` is an old `got` property which we like to get rid of.

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,55 @@
+
+  protected handleGenericErrors(err: HttpError): never {
+    this.handleSpecificErrors(err);
+    if (err.statusCode !== undefined) {
+      if (
+        err.statusCode === 429 ||
+        (err.statusCode >= 500 && err.statusCode < 600)
+      ) {
+        throw new ExternalHostError(err);
+      }
+    }
```
```suggestion
    if (err.response?.statusCode !== undefined) {
      if (
        err.response?.statusCode === 429 ||
        (err.response?.statusCode >= 500 && err.response?.statusCode < 600)
      ) {
        throw new ExternalHostError(err);
      }
    }
```

`statusCode` is a legacy `got` property, should no longer be used.

viceice commented on lib/datasource/index.ts:
```diff
@@ -9,6 +9,8 @@ import { clone } from '../util/clone';
 import { regEx } from '../util/regex';
 import * as allVersioning from '../versioning';
 import datasources from './api';
+import { CdnJs } from './cdnjs';
+import { Clojure } from './clojure';
```
```suggestion
```

Not needed, see below

viceice commented on lib/util/http/index.ts:
```diff
@@ -47,6 +47,10 @@ export interface HttpResponse<T = string> {
   authorization?: boolean;
 }
 
+export interface HttpError extends Error {
+  statusCode: number;
+}
```
Use `HttpError` from `lib/util/types.ts`
```suggestion
```

See other comments for why not use `statusCode` 😉

viceice commented on lib/versioning/loose/generic.ts:
```diff
@@ -189,7 +189,6 @@ export abstract class GenericVersioningApi<
     return versions.find((v) => this.equals(v, range)) || null;
   }
 
-  // eslint-disable-next-line class-methods-use-this
```
revert, as you like to revert the eslint setting?

viceice commented on package.json:
```diff
@@ -7,8 +7,8 @@
   "scripts": {
-    "build": "run-s clean generate:* compile:*",
-    "clean": "rimraf dist",
+    "build": "run-s clean prepare:* generate:* compile:*",
+    "clean": "rimraf dist tmp",
```
Wrong merge?

viceice commented on tools/jest-gh-reporter.ts:
```diff
@@ -45,7 +45,6 @@ function getPos(msg: string): Record<string, string> {
 }
 
 class GitHubReporter extends BaseReporter {
-  // eslint-disable-next-line class-methods-use-this
```
wrong?

JamieMagee commented on lib/datasource/cdnjs/index.ts:
```diff
@@ -24,38 +22,43 @@ interface CdnjsResponse {
 
-export async function getReleases({
-  lookupName,
-  registryUrl,
```
Yes. That's the trouble with long-lived branches :cry:

JamieMagee commented on .eslintrc.js:
```diff
@@ -42,6 +42,9 @@ module.exports = {
     'no-underscore-dangle': 0,
     'no-negated-condition': 'error',
+    // TODO: re-enable once datasource class migration
+    'class-methods-use-this': 0,
```
Sure:

`getReleases` in `clojure` doesn't use `this` as it uses `getReleases` from `maven` instead of using its own implementation.

`handleSpecificErrors` in the abstract class `Datasource` is a no-op, and not marked as abstract, as not all datasources may need to handle specific errors, so `handleSpecificErrors` is flagged.

I figured it was easier to disable this here during migration, and it can then be re-enabled for specific methods later? I can flag those specific methods now, if you think it's easier or better.

viceice replied:
i think it better to flag the causing code, otherwise you need a lot effort to fix it again

JamieMagee replied:
Okay, removed this from here, and added it to each individual instance.

viceice commented on lib/manager/deps-edn/extract.ts:
```diff
@@ -16,7 +16,7 @@ export function extractPackageFile(content: string): PackageFile {
 
     deps.push({
-      datasource: datasourceClojure.id,
+      datasource: new ClojureDatasource().id,
```
we should export the id also directly, so we do not need those changes 😕 Looks poretty bad to instanciate for id property only

JamieMagee replied:
I _think_ I have a better solution for this. I define `Datasource` as

```ts
export abstract class Datasource implements DatasourceApi {
  protected constructor(id: string) {
    this.http = new Http(id);
    this.id = id;
  }
}
```

and implement it like so:

```ts
export class CdnJsDatasource extends Datasource {
  static readonly id = 'cdnjs';

  constructor() {
    super(CdnJsDatasource.id);
  }
}
```

The main issue is this doesn't actually require `id` to be defined. I could put an assert in the constructor that would be caught in unit tests, but throwing exceptions in constructors is generally considered bad practice.

I can't think of a good way to get the compiler to enforce this constraint.

viceice replied:
Would need strict null check in typescript. But that would need a lot of work. 😉

viceice replied:
Simpler:

```ts
export abstract class Datasource implements DatasourceApi {
  protected http: Http;
  protected constructor(protected readonly id: string) {
    this.http = new Http(id);
  }
}
```

viceice commented on lib/datasource/index.spec.ts:
```diff
@@ -44,9 +44,20 @@ describe(getName(__filename), () => {
       }
       return true;
     }
+    function filterClassBasedDatasources(name: string): boolean {
+      return !['cdnjs', 'clojure'].includes(name);
+    }
```
better use `getDatasources()` and check `obj instanceOf Datasource`, so we do not need to change the test on every datasource migration.

JamieMagee replied:
Will come back to this later.

I can't directly modify `datasource.getDatasources()` and `clone` is returning `{}` for some reason 🤔

viceice replied:
clone with `new Map(datasource.getDatasources())` ? 😏

viceice reviewed (commented):
I think we should only convert this two datasources with this pr and convert others in separate pr's, so it's easier to review

viceice commented on .eslintrc.js:
```diff
@@ -42,6 +42,7 @@ module.exports = {
     'prefer-template': 0,
     'no-underscore-dangle': 0,
     'no-negated-condition': 'error',
+
```
😏

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,51 @@
+} from './types';
+
+export abstract class Datasource implements DatasourceApi {
+  protected constructor(public readonly id: string) {
+    this.http = new Http(id);
+  }
+
+  caching: boolean;
+
+  defaultConfig: Record<string, unknown>;
+
+  customRegistrySupport: boolean;
+
+  defaultRegistryUrls: string[];
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge';
+
+  protected http: Http;
+
+  abstract getReleases(
+    getReleasesConfig: GetReleasesConfig
+  ): Promise<ReleaseResult | null>;
+
+  getDigest?(config: DigestConfig, newValue?: string): Promise<string>;
+
+  // eslint-disable-next-line class-methods-use-this
+  handleSpecificErrors(err: HttpError): void {}
```
We should define an order to we chave consistency. 🤔

JamieMagee replied:
An order for what?

viceice replied:
member type order and alphabetic naming. maybe we need to open an issue to discuss best practices for us?

JamieMagee replied:
Not sure it's that big a deal right now. Maybe later if this becomes more busy.

For now I've sorted alphabetically, with `protected http` at the bottom as it's a different modifier.

viceice replied:
you are right, I've opened an issue to track this, so we can make a desision in future. ref #9948

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,51 @@
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge';
```
maybe add default value?

JamieMagee replied:
Default should be `first`?

viceice replied:
yes, looks like most registries use `first`, so let's configure as default.

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,51 @@
+
+  defaultConfig: Record<string, unknown>;
+
+  customRegistrySupport: boolean;
```
defaults to false?

JamieMagee replied:
Matches the current `DatasourceApi`. That's correct?

viceice replied:
a quick seach shows this is truw for most datasources, so let's make this true by default, so only some datasources need to override :)

viceice commented on lib/datasource/index.spec.ts:
```diff
@@ -34,19 +34,27 @@ describe(getName(), () => {
+    function filterClassBasedDatasources(name: string): boolean {
+      return !['cdnjs', 'clojure'].includes(name);
     }
     const dss = datasource.getDatasources();
```
dss is a `Map`, so clone like this:

```suggestion
    const dss = new Map(datasource.getDatasources());
```

viceice replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/Map

viceice reviewed (commented):
needs lint fix 😉

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,51 @@
+        err.response?.statusCode === 429 ||
+        (err.response?.statusCode >= 500 && err.response?.statusCode < 600)
+      ) {
+        throw new ExternalHostError(err);
```
needs coverage 😉

JamieMagee replied:
It turns out that neither of the 2 datasources I converted actually use this logic. I think either `cargo` must have used it (before I dropped it due to conflicts) or I took it as a decent starting point for generic datasource error handling.

viceice commented on lib/datasource/datasource.ts:
```diff
@@ -0,0 +1,51 @@
+    this.http = new Http(id);
+  }
+
+  caching: boolean;
+
+  customRegistrySupport = true;
+
+  defaultConfig: Record<string, unknown>;
+
+  defaultRegistryUrls: string[];
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge' = 'first';
```
```suggestion
  caching: boolean | undefined;

  customRegistrySupport = true;

  defaultConfig: Record<string, unknown> | undefined;

  defaultRegistryUrls: string[] | undefined;

  defaultVersioning: string | undefined;

  registryStrategy: 'first' | 'hunt' | 'merge' | undefined = 'first';
```

They should stay optional and we want to make future prove for typescript `strictNullCheck`.

JamieMagee commented:
@viceice I think I've addressed all your comments.

@rarkins I think this is ready for your review.

JamieMagee commented:
Pls no more conflicts 😨 🤞

viceice commented on lib/datasource/datasource.spec.ts:
```diff
@@ -0,0 +1,45 @@
+import { getName } from '../../test/util';
+import { EXTERNAL_HOST_ERROR } from '../constants/error-messages';
+import { Datasource } from './datasource';
+import { GetReleasesConfig, ReleaseResult } from './types';
```
```suggestion
import type { GetReleasesConfig, ReleaseResult } from './types';
```

viceice commented on lib/datasource/clojure/index.spec.ts:
```diff
@@ -0,0 +1,265 @@
+import { getName, loadFixture, mocked } from '../../../test/util';
+import * as hostRules from '../../util/host-rules';
+import { id as versioning } from '../../versioning/maven';
+import { ClojureDatasource as datasource } from '.';
```
Why renaming? Makes code harder to read.

viceice commented on lib/datasource/datasource.spec.ts:
```diff
@@ -0,0 +1,45 @@
+    super('test');
+  }
+
+  // eslint-disable-next-line class-methods-use-this
```
```suggestion
```

we are using `this` 😏

viceice reviewed (commented):
Just some small stuff

viceice reviewed (approved):
🥳🥳
